### PR TITLE
WIP init card10 support

### PR DESCRIPTION
**Pull request by xermicus**
It works already kind of but there are many TODOs / quick hacks to make it work properly.

To build you need the firmware and the [card10 toolchain](https://firmware.card10.badge.events.ccc.de/how-to-build.html); export FIRMWARE_HOME to point to the card10 firmware dir.

FixMe:
- [x] Tasks are not needed since the l0dable gets only 1 core anyways
- [x] [WONTFIX] Make `-mfloat-abi=hard` possible (there seems to be some incompatible BLE stuff inside card10 firmware that prevents it)
- [x] Make random_seed() work
- [x] Fix x/y axis
- [x] Make wait_until() / wait_until_break() work properly
- [x] Make clear() work properly
- [x] Implement get()
- [x] Add binary to [badge hatchery](https://badge.team/?badge=card10)

### CODE REVIEW

File: GNUmakefile [hunk 1]
```diff
@@ -162,6 +162,11 @@ ML_OBJECTS := $(ML_SOURCES:.c=.o)
 
 # --- Include other makefiles ---
 include Makefiles/3ds.GNUmakefile
+<<<<<<< HEAD
```
Comment (vifino) [inline GNUmakefile]: Merge conflict.

File: Makefiles/card10.GNUmakefile [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+CC := arm-none-eabi-gcc
+AR := arm-none-eabi-ar
+LD := arm-none-eabi-ld
+STRIP := arm-none-eabi-strip
+
+# CFLAGS
+# various
+#CFLAGS += -mthumb
+#CFLAGS += -mcpu=cortex-m4
```
Comment (vifino) [inline Makefiles/card10.GNUmakefile]: Why is this commented out?

File: src/modules/out_card10.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#if !defined(MATRIX_ORDER_PLAIN) && !defined(MATRIX_ORDER_SNAKE)
+#define MATRIX_ORDER_PLAIN
+#endif
+
+#ifndef MATRIX_X
+#error Define MATRIX_X as the matrixes X size.
+#endif
+
+#ifndef MATRIX_Y
+#error Define MATRIX_Y as the matrixes Y size.
```
Comment (vifino) [inline src/modules/out_card10.c]: probably makes sense to hardcode those values

File: src/modules/out_card10.c [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	// 5 4 3
+	// 6 7 8
+	return (((y % 2) == 0 ? x : (MATRIX_X - 1) - x) + MATRIX_X*y);
+#endif
+}
+
+RGB get(int _modno, int x, int y) {
+	//size_t px = ppos(x, y);
+	// TODO
+	return RGB(0,0,0);
```
Comment (vifino) [inline src/modules/out_card10.c]: shouldn't be hard to implement, probably'll do it myself.

File: src/modules/out_card10.c [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	uint16_t converted = RGB2RGB565(color);
+	fb.fb[y][x][0] = converted >> 8;
+	fb.fb[y][x][1] = converted & 0xFF;
+	return 0;
+}
+
+
+int clear(void) {
+	// TODO this does not work yet
+	return epic_disp_clear(RGB2RGB565(RGB(0,0,0)));
```
Comment (vifino) [inline src/modules/out_card10.c]: i don't think there is a need to convert 0, 0 is always 0

File: src/modules/out_card10.c [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+int render(void) {
+	return epic_disp_framebuffer(&fb);
+}
+
+ulong wait_until(ulong desired_usec) {
+	return;
+	//return timers_wait_until_core(desired_usec);
```
Comment (vifino) [inline src/modules/out_card10.c]: why is this commented out?

File: src/os/os_freertos.c [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+	printf("Signaling.\n");
+	xSemaphoreGive(ev);
+}
+
+void oscore_event_free(oscore_event ev) {
+	vSemaphoreDelete(ev);
+}
+
+// Time keeping.
+// Since we don't have a RTC, presumably,
```
Comment (vifino) [inline src/os/os_freertos.c]: we have an rtc, might make sense to implement the rtc here

File: Makefiles/card10.GNUmakefile (hunk 2 above)
Comment (xermicus) [reply Makefiles/card10.GNUmakefile]: it's in sledconf

File: src/modules/out_card10.c (hunk 4 above)
Comment (xermicus) [reply src/modules/out_card10.c]: :+1:

File: src/modules/out_card10.c (hunk 5 above)
Comment (xermicus) [reply src/modules/out_card10.c]: lol yeah

File: src/modules/out_card10.c (hunk 6 above)
Comment (xermicus) [reply src/modules/out_card10.c]: it does not work and I dont have a debugging kit

File: src/os/os_freertos.c (hunk 7 above)
Comment (xermicus) [reply src/os/os_freertos.c]: :+1:

**Comment by xermicus** [comment]
you're welcome :-) 

yeah I just did it in a different branch because it's another cpu arch
it was also a nice test to see how portable the os_freertos already is

File: Makefiles/card10.GNUmakefile (hunk 2 above)
Comment (vifino) [reply Makefiles/card10.GNUmakefile]: maybe we should make it add those for the card10 anyway, we know what the hardware looks like so it's fine.

File: src/modules/out_card10.c (hunk 6 above)
Comment (vifino) [reply src/modules/out_card10.c]: ah, i'll try to assemble mine

File: src/modules/out_card10.c [hunk 8]
```diff
@@ -41,16 +48,16 @@ int ppos(int x, int y) {
...
 	return RGB5652RGB(converted);
 }
 
 int set(int modno, int x, int y, RGB color) {
 	// No OOB check, because performance.
 	uint16_t converted = RGB2RGB565(color);
-	fb.fb[LCD_Y-1-y][x][0] = converted >> 8;
-	fb.fb[LCD_Y-1-y][x][1] = converted & 0xFF;
+	fb.fb[LCD_Y-1-y][LCD_X-x][0] = converted >> 8;
+	fb.fb[LCD_Y-1-y][LCD_X-x][1] = converted & 0xFF;
```
Comment (vifino) [inline src/modules/out_card10.c]: Ah, so both axis needed flipping?

Comment (xermicus) [reply src/modules/out_card10.c]: apparently yes

**Comment by xermicus** [comment]
All FixMes are done :-) Also added some performance compiler flags that make a visible difference

**Comment by vifino** [comment]
Looks good, cleanup can be done in master.

**Pull request merged**